Clarify upload docs (#944)
huggingface/huggingface_hub: docs/source/how-to-upstream.mdx

# Upload files to the Hub

Sharing your files and work is an important aspect of the Hub. The `huggingface_hub` offers several options for uploading your files to the Hub. You can use these functions independently or integrate them into your library, making it more convenient for your users to interact with the Hub. This guide will show you how to push files:

* without using Git.
* that are very large with [Git LFS](https://git-lfs.github.com/).

> * with with the `commit` context manager.

Reviewer: ```suggestion
* with the `commit` context manager.
```

# Upload files to the Hub

> Sharing your files and work is an important aspect of the Hub. The `huggingface_hub` offers several options for uploading your files to the Hub. You can use these functions independently or integrate them into your library, making it more convenient for your users to interact with the Hub. This guide will show you how to:

Reviewer 2: ```suggestion
Sharing your files and work is an important aspect of the Hub. The `huggingface_hub` offers several options for uploading your files to the Hub. You can use these functions independently or integrate them into your library, making it more convenient for your users to interact with the Hub. This guide will show you how to push files:
```

> * Push files without Git.
> * Push very large files with [Git LFS](https://git-lfs.github.com/).
> * Push files with the `commit` context manager.

Reviewer 2: ```suggestion
* without using Git
* that are very large files with [Git LFS](https://git-lfs.github.com/).
* using the `commit` context manager.
```

* Push files with the [`~Repository.push_to_hub`] function.

Whenever you want to upload files to the Hub, you need to log in to your Hugging Face account:

- Log in to your Hugging Face account with the following command:

   ```bash
   huggingface-cli login
   ```

- Alternatively, if you prefer working from a Jupyter or Colaboratory notebook, log in with [`notebook_login`]:

   ```python
   >>> from huggingface_hub import notebook_login
   >>> notebook_login()
   ```

   [`notebook_login`] launches a widget in your notebook from which you can enter your Hugging Face credentials.

## Push files without Git

> If you don't have Git installed on your system, use [`create_commit`] to push your files to the Hub. [`create_commit`] uses the HTTP protocol to upload files to the Hub. Any large or binary files are automatically uploaded with Git LFS. 

Reviewer 2: I would not assume people know what Git LFS is, so maybe we can remove the last sentence or move it to a tip.

For example, if you want to upload two files and delete a file in a Hub repository:

1. Use the appropriate `CommitOperation` to add and delete a file:

```py
>>> from huggingface_hub import HfApi, CommitOperationAdd, CommitOperationDelete
>>> api = HfApi()
>>> operations = [
...     CommitOperationAdd(path_in_repo="LICENSE.md", path_or_fileobj="~/repo/LICENSE.md"),
...     CommitOperationAdd(path_in_repo="weights.h5", path_or_fileobj="~/repo/weights-final.h5"),
...     CommitOperationDelete(path_in_repo="old-weights.h5"),
... ]
```

2. Pass your operations to [`create_commit`]:

```py
>>> api.create_commit(
...     repo_id="lysandre/test-model",
...     operations=operations,
...     commit_message="Upload my model weights and license",
... )
```

In addition to [`upload_file`] and [`upload_folder`], the following functions also use [`create_commit`] under the hood: 

* [`delete_file`] deletes a single file from a repository on the Hub. 
* [`metadata_update`] updates a repository's metadata.

For more detailed information, take a look at the [`HfApi`] reference.

## Push files with Git LFS

> For large files (>5GB), you need to install a custom transfer agent for Git LFS:

Reviewer 2: All files with 10mb or more also use Git LFS, just that the `lfs-enable-largefiles` is needed for very very large files. Maybe mention that all methods will automatically add your files to git lfs if needed, and just use lfs-enable-largefiles for these very large files

   ```bash
   huggingface-cli login
   ```

- Alternatively, if you prefer working from a Jupyter or Colaboratory notebook, login with [`notebook_login`]:

   ```python
   >>> from huggingface_hub import notebook_login
   >>> notebook_login()
   ```

   [`notebook_login`] launches a widget in your notebook from which you can enter your Hugging Face credentials.

## Upload a file

Once you've created a repository with the [`create_repo`](./how-to-manage#create-a-repository) function, you can upload your files to your repository. Use the [`upload_file`] function to upload a file. 

Specify the path of the file to upload, where you want to upload the file to in the repository, and the name of the repository you want to add the file to. Depending on your repository type, you can optionally set the repository type as a `dataset`, `model` or `space`.

```py
>>> from huggingface_hub import HfApi
>>> api = HfApi()
>>> api.upload_file(path_or_fileobj="/path/to/local/folder/README.md", 
...                 path_in_repo="README.md", 
...                 repo_id="username/test-dataset",
...                 repo_type="dataset",
... )
```

## Upload a folder

> Use the [`upload_folder`] function to upload a local folder to a repository. Specify the path of the local folder to upload, where you want to upload the folder to in the repository, and the name of the repository you want to add the folder to. Depending on your repository type, you can optionally set the repository type as a `dataset`, `model` or `space`.

Reviewer: Similar to "upload a file", I would add here that you first need to create a repo before using this function

> Use the [`upload_folder`] function to upload a local folder to a repository. Specify the path of the local folder to upload, where you want to upload the folder to in the repository, and the name of the repository you want to add the folder to. Depending on your repository type, you can optionally set the repository type as a `dataset`, `model` or `space`.

Reviewer: ```suggestion
Use the [`upload_folder`] function to upload a local folder to an existing repository. Specify the path of the local folder to upload, where you want to upload the folder to in the repository, and the name of the repository you want to add the folder to. Depending on your repository type, you can optionally set the repository type as a `dataset`, `model` or `space`.
```
Similar to "upload a file", would be nice to clarify here that a repo first needs to be created

   [`notebook_login`] will launch a widget in your notebook from which you can enter your Hugging Face credentials.

## Upload a file

> Use the [`upload_file`] function to upload a file to a repository. Let the function know where the file you want to upload is, where you want to upload the file to in the repository, and the name of the repository you want to add the file to.

Reviewer: Would be great to clarify here whether the repo_id needs to already exist on the hub, or whether a repo will be created automatically in case it doesn't exist already.

Cause personally I don't know actually :D

Reviewer 2: For creating a repository they should likely use `create_repo`. `Repository` is git-based and would clone the repo locally. `create_repo` + `upload_file` take care of it.

Reviewer 3: Yes, `Repository` shouldn't be used to create the repository. It can create it, but this feature is likely to disappear in favor of the unambiguous `create_repo` instead (see discussions in https://github.com/huggingface/huggingface_hub/issues/672)

```py
>>> from huggingface_hub import HfApi
>>> api = HfApi()
>>> api.upload_file(path_or_fileobj="/home/lysandre/dummy-test/README.md", 
...                 path_in_repo="README.md", 
...                 repo_id="lysandre/test-model",
... )
```

## Upload a folder

Use the [`upload_folder`] function to upload a local folder to a repository. Specify the path of the local folder to upload, where you want to upload the folder to in the repository, and the name of the repository you want to add the folder to.

```py
>>> from huggingface_hub import HfApi
>>> api = HfApi()
>>> api.upload_folder(folder_path="/path/to/local/folder",
...                   path_in_repo="my-dataset/train",
...                   repo_id="username/test-dataset"

> ...                   repo_type="datasets",

Reviewer: In the code snippet above, `repo_type` isn't used, but here it is.

I'd say either use it for both (and explain what it does), or for neither.

```python
>>> from huggingface_hub import Repository

> >>> repo = Repository(local_dir="path/to/local/repo")

Reviewer: Here, I don't understand why you specify a local directory and then do a git pull.

Why would you need to git pull in case there's no remote specified?

Reviewer: And again here, would be useful to explain whether a remote repository needs to exist already, or whether it will be created automatically in case it doesn't.

Author: Hmm, my understanding of `Repository` is that if I have an existing local clone, I can just initialize it from the `local_dir`. Then I can `git_pull` to update it with the repository on the Hub.

Let me know if I am misunderstanding how that works! :)

Reviewer: Ok in that case I would add a comment to clarify this, maybe replace `path/to/local/repo` with `path/to/local/clone` in the code snippet above

Reviewer: Oh nvm, just saw you added additional text above the code snippet. Fine for me!

You should install this for each model repository that contains a model file. Once installed, you are now able to push files larger than 5GB.

## Upload without Git installed

> If you don't have Git installed on your system, use the [`create_commit`] function to upload your file. [`create_commit`] uses the HTTP protocol to upload files to the Hub. Large files and binary files are automatically uploaded with the Git LFS protocol.

Author: Good point! I'll add a tip for users to use one of the other functions before reaching for `create_commit` :)

You should install this for each model repository that contains a model file. Once installed, you are now able to push files larger than 5GB.

## Upload without Git installed

<Tip>

The [`create_commit`] is really useful for certain use-cases. In general though, it is simpler to use one of the other uploading functions mentioned above.

</Tip>

> If you don't have Git installed on your system, use the [`create_commit`] function to upload your file. [`create_commit`] uses the HTTP protocol to upload files to the Hub. Large files and binary files are automatically uploaded with the Git LFS protocol.

Reviewer: The upload_file and upload_folder functions also just use HTTP and don't require git. From their docs:

> The upload is done through a HTTP post request, and doesn't require git or git-lfs to be installed.

So I'd say here that create_commit is just a low-level API, in case you want to work at a commit level.
